tools/infrasys/src/error.rs

```diff
@@ -0,0 +1,160 @@
+    },
+
+    #[snafu(display(
+        "Failed to find field {} after attempting to create resource {}",
+        what,
+        resource_name
+    ))]
+    ParseResponse { what: String, resource_name: String },
+
+    #[snafu(display("Failed to get bucket policy statement for bucket {}", bucket_name))]
+    GetPolicyStatement { bucket_name: String },
+
+    #[snafu(display(
+        "Failed to update bucket policy for bucket {}: {}",
+        bucket_name,
+        source
+    ))]
+    PutPolicy {
+        bucket_name: String,
+        source: rusoto_core::RusotoError<rusoto_s3::PutBucketPolicyError>,
+    },
+
+    #[snafu(display("Failed to push object to bucket {}: {}", bucket_name, source))]
+    PutObject {
+        bucket_name: String,
+        source: rusoto_core::RusotoError<rusoto_s3::PutObjectError>,
+    },
+
+    #[snafu(display("Missing environment variable '{}'", var))]
+    Environment {
+        var: String,
+        source: std::env::VarError,
+    },
+
+    #[snafu(display("Failed to read file at {}: {}", path.display(), source))]
+    FileRead { path: PathBuf, source: io::Error },
+
+    #[snafu(display("Failed to open file at {}: {}", path.display(), source))]
+    FileOpen { path: PathBuf, source: io::Error },
+
+    #[snafu(display("Failed to write file in {}: {}", path.display(), source))]
+    FileWrite { path: PathBuf, source: io::Error },
+
+    #[snafu(display("File already exists at {}", path.display()))]
+    FileExists { path: PathBuf },
+
+    #[snafu(display(
+        "Failed to create keys due to invalid key config. Missing {}.",
+        missing
+    ))]
+    KeyConfig { missing: String },
+
+    #[snafu(display(
+        "Failed to create new keys or access pre-existing keys in available_keys list."
+    ))]
+    KeyCreation,
+
+    #[snafu(display("Invalid path {} for {}", path.display(), thing))]
+    Path { path: PathBuf, thing: String },
+
+    #[snafu(display("Failed to create directory {}: {}", path.display(), source))]
+    Mkdir { path: PathBuf, source: io::Error },
+
+    #[snafu(display("Failed to convert {} to URL: {}", input, source))]
+    ParseUrl {
+        input: String,
+        source: url::ParseError,
+    },
+
+    #[snafu(display("Failed to parse {} to int: {}", what, source))]
+    ParseInt {
+        what: String,
+        source: std::num::ParseIntError,
+    },
+
+    #[snafu(display("Publication/Root key threshold must be <= {}, currently {}", num_keys.to_string(), threshold))]
+    InvalidThreshold { threshold: String, num_keys: usize },
+
+    #[snafu(display("Failed to convert updated Infra.toml information to yaml: {}", source))]
+    InvalidYaml { source: serde_yaml::Error },
+
+    #[snafu(display("Failed to convert {} to yaml: {}", what, source))]
+    InvalidJson {
+        what: String,
+        source: serde_json::Error,
+    },
+
+    #[snafu(display("Failed to get parent of path: {}", path.display()))]
+    Parent { path: PathBuf },
```

Comment:
If I were to switch all Parent to Path errors, I would probably do something like:
```
toml_path
            .parent()
            .context(error::Path { path: toml_path.parent().unwrap() , thing: "Infra.toml parent" })?
            .join("Infra.lock"),
```

Are we okay with unwrapping in this case?
I could also list the path as the original toml_path, but the failure message might be a bit misleading?

Reply:
I think having a separate error variant for this makes sense. The path itself isn't necessarily invalid, it's just that it doesn't have a parent (e.g. it might be `/`).